system/jlib/jptree.cpp

```diff
@@ -9481,3 +9542,341 @@ void copyPropIfMissing(IPropertyTree & target, const char * targetName, IPropert
+            }
+            break;
+    }
+    ch += chlen;
+    remaining -= chlen;
+    if (first)
+        first = false;
+    return s;
+}
+
+StringBuffer &encodePTreeName(StringBuffer &s, unsigned size, const char *value, const char *startEncoding)
+{
+    if (!value || !size)
+        return s.append("_0");
+    bool isattr = ('@'==*value);
+    if (isattr && size==1)
+        return s.append("@_0");
+
+    if (startEncoding)
+    {
+        //see if we have any existing escape characters to expand ('_' ==> "__")
+        void *realStart = memchr((void *)value, '_', (size_t)(startEncoding - value));
+        if (realStart)
+            startEncoding = (const char *) realStart;
+        unsigned skipSize = startEncoding-value;
+        if (skipSize > size)
+            return s;
+        s.append(skipSize, value);
+        value = startEncoding;
+        size -= skipSize;
+    }
+    bool hexsequence = false;
+    bool first = true;
+    while (size)
+        encodePTreeNameUtf8Char(s, value, size, isattr, hexsequence, first);
+    checkEndHexSequence(s, hexsequence);
+    return s;
+}
+
+StringBuffer &encodePTreeName(StringBuffer &s, const char *value, const char *startEncoding)
+{
+    if (!value)
+        return s.append("_0");
+    return encodePTreeName(s, strlen(value), value, startEncoding);
+}
+
+//Only encode the PTREE XPATH name if necessary. That matches the way internal PTREE names are stored allowing the resulting XPATH to be used directly
+StringBuffer &appendPTreeXPathName(StringBuffer &s, unsigned size, const char *value)
+{
+    if (!size || !value)
+        return s.append("_0");
+    const char *startEncoding = findFirstInvalidPTreeNameChar(value, size);
+    if (!startEncoding)
+        return s.append(value);
+    return encodePTreeName(s, size, value, startEncoding);
+}
+
+StringBuffer &appendPTreeXPathName(StringBuffer &s, const char *value)
+{
+    if (isEmptyString(value))
+        return s;
+    return appendPTreeXPathName(s, strlen(value), value);
+}
+
+void markPTreeNameEncoded(IPropertyTree *tree)
+{
+    if (!tree)
+        return;
+    PTree *pt = static_cast<PTree*>(tree);
+    pt->markNameEncoded();
+}
+
+bool isPTreeNameEncoded(const IPropertyTree *tree)
+{
+    if (!tree)
+        return false;
+    const PTree *pt = static_cast<const PTree*>(tree);
+    return pt->isNameEncoded();
+}
+
+void setPTreeAttribute(IPropertyTree *tree, const char *name, const char *value, bool markEncoded)
+{
+    if (!tree || isEmptyString(name))
+        return;
+    PTree *pt = static_cast<PTree*>(tree);
+    pt->setAttribute(name, value, markEncoded);
+}
+
+bool isPTreeAttributeNameEncoded(const IPropertyTree *tree, const char *name)
+{
+    if (!tree || isEmptyString(name))
+        return false;
+    const PTree *pt = static_cast<const PTree*>(tree);
+    return pt->isAttributeNameEncoded(name);
+}
+
+static void decodePTreeNameHexEncoding(StringBuffer &s, const char *&input, unsigned &remaining)
+{
+    byte ch = 0;
```

Comment:
I'm not sure how good code optimizations are these days.  But I think you might want to assign input, remaining to local variables at the head of the function, and assign them to the reference parameters as the end. Otherwise all references to input be indirect.

Reply:
It looks like the whole set of functions are inlined anyway